### PR TITLE
[WIP] Serving API definition and controller implementation.

### PR DESCRIPTION
**Pull request by SimonCqk**
work in progress and do-not-merge.

### CODE REVIEW

File: apis/serving/v1alpha1/inference_types.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	// Batching specify batching strategy to control how to forward a batch of
+	// aggregated inference requests to backend service.
+	Batching *BatchingStrategy `json:"batching,omitempty"`
+}
+
+type PredictorStatus struct {
+	// Name is the name of current predictor.
+	Name string `json:"name"`
+	// Replicas is the expected replicas of current predictor.
+	Replicas int32 `json:"replicas"`
```
Comment (jian-he) [inline apis/serving/v1alpha1/inference_types.go:84]: this is already specifed in Spec, do we still need it here ?

File: controllers/serving/inference_controller.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+		vsvcInCluster.Spec = *vsvc.Spec.DeepCopy()
+		return ir.client.Update(context.Background(), &vsvcInCluster)
+	}
+
+	return nil
+}
+
+// syncServiceForInference sync an entrypoint service for inference service, it provides user with
+// a unified host to access inference service, thought the inference service may serves more than one
+// model version at a time, the backward network topology will handle traffic routing.
```
Comment (jian-he) [inline controllers/serving/inference_controller.go:234]: typo: thought, backward

File: controllers/serving/predictor.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	} else if !errors.IsNotFound(err) {
+		return err
+	}
+
+	// At this point, service not found and create a new one.
+
+	httpPort := getContainerPort(predictor, v1alpha1.DefaultPredictorContainerHttpPortName, v1alpha1.DefaultPredictorContainerHttpPort)
+	grpcPort := getContainerPort(predictor, v1alpha1.DefaultPredictorContainerGrpcPortName, v1alpha1.DefaultPredictorContainerGrpcPort)
+
+	svc = corev1.Service{
```
Comment (jian-he) [inline controllers/serving/predictor.go:124]: what's the different between this service created vs the one created at inference_controller.go reconcile loop?

File: apis/serving/v1alpha1/inference_types.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	// ModelVersion specifies the name of target model-with-version to be loaded
+	// in serving service, ModelVersion object has to be created before serving
+	// service deployed.
+	ModelVersion string `json:"modelVersion"`
+
+	// Replicas specify the expected predictor replicas.
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// CanaryTrafficPercent defines the traffic split percentage.
+	CanaryTrafficPercent *int32 `json:"canaryTrafficPercent,omitempty"`
```
Comment (jian-he) [inline apis/serving/v1alpha1/inference_types.go:67]: rename to trafficPercent

File: apis/serving/v1alpha1/inference_types.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+type PredictorStatus struct {
+	// Name is the name of current predictor.
+	Name string `json:"name"`
+	// Replicas is the expected replicas of current predictor.
+	Replicas int32 `json:"replicas"`
+	// ReadyReplicas is the ready replicas of current predictor.
+	ReadyReplicas int32 `json:"readyReplicas"`
+	// TrafficPercent is the traffic distribution routed to this service.
+	TrafficPercent *int32 `json:"trafficPercent,omitempty"`
```
Comment (jian-he) [inline apis/serving/v1alpha1/inference_types.go:88]: this TrafficPercent is same as in spec ? can remove

File: apis/serving/v1alpha1/inference_types.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	// Endpoint is the available endpoints of current predictor.
+	Endpoint string `json:"endpoint"`
```
Comment (jian-he) [inline apis/serving/v1alpha1/inference_types.go:90]: add TODO  for the fields that are not yet supported

File: apis/serving/v1alpha1/inference_types.go (hunk 1 above)
Comment (SimonCqk) [reply apis/serving/v1alpha1/inference_types.go:84]: `spec.replicas` represents expected number of replicas, `status.replicas` represents number of replicas have been created

File: controllers/serving/predictor.go (hunk 3 above)
Comment (SimonCqk) [reply controllers/serving/predictor.go:124]: there comes two-level of services in design:
1. `service` for inference-service: main entry of user traffic, host for users to access inference service.
2. `services` for each predictor: entry for each predictor deployment.

hence a user request will be routed as: 
```
requests ---> service(host: inference.default.svc)
                             |---> service1 (host: inference-predictor-model-1-1.default.svc)---> deployment 1.1
                             |---> service2 (host: inference-predictor-model-1-2.default.svc)---> deployment 1.2
```

File: apis/serving/v1alpha1/inference_types.go (hunk 4 above)
Comment (SimonCqk) [reply apis/serving/v1alpha1/inference_types.go:67]: done

File: apis/serving/v1alpha1/inference_types.go (hunk 5 above)
Comment (SimonCqk) [reply apis/serving/v1alpha1/inference_types.go:88]: `spec.trafficPercent` represents the expected percentage while `status.trafficPercent` represents the percentage applied

File: apis/serving/v1alpha1/inference_types.go (hunk 1 above)
Comment (jian-he) [reply apis/serving/v1alpha1/inference_types.go:84]: then, the comment is wrong.

File: config/crd/patches/cainjection_in_servingservices.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
+# CRD conversion requires k8s 1.13 or later.
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: $(CERTIFICATE_NAMESPACE)/$(CERTIFICATE_NAME)
+  name: servingservices.serving.kubedl.io
```
Comment (jian-he) [inline config/crd/patches/cainjection_in_servingservices.yaml:8]: the name need to be changed

File: apis/serving/v1alpha1/inference_types.go (hunk 5 above)
Comment (jian-he) [reply apis/serving/v1alpha1/inference_types.go:88]: 看起来不对？status里面才是真正的percentage，spec里面是weight ?

File: config/crd/patches/cainjection_in_servingservices.yaml (hunk 7 above)
Comment (SimonCqk) [reply config/crd/patches/cainjection_in_servingservices.yaml:8]: removed

File: controllers/model/modelversion_controller.go [hunk 8]
```diff
@@ -139,6 +139,11 @@ func (r *ModelVersionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 			modelVersionStatus.Image = modelVersion.Spec.ImageRepo + ":v" + versionId
 			modelVersionStatus.ImageBuildPhase = modelv1alpha1.ImageBuilding
 			modelVersionStatus.Message = "Image building started."
+			imgBuildDockerfile := createImageBuildDockerfile(modelVersion.Namespace)
```
Comment (jian-he) [inline controllers/model/modelversion_controller.go:142]: The configMap is the same for all. But invocation here could cause creation many times

**Comment by jian-he** [comment]
two follow up issues;
- TrafficPercent in Percent can be renamed to TrafficWeight, as this is not a percentage.
- "c.Command = []string{"/bin/sh", "-c", "cp", v1alpha1.DefaultModelInImagePath, destModelPath}" 
This may need to be "cp -R" , as it can be a directory ? 
and DefaultModelInImagePath can be changed from "/model" to "/model/" , to do recursive copy under the directory

**Comment by jian-he** [comment]
merge this in first

**Pull request merged**